imagebuildah/build.go

```diff
@@ -675,42 +693,304 @@ func (b *Executor) Execute(ib *imagebuilder.Builder, node *parser.Node) error {
+		}
+		currNode = currNode.Next
+	}
+	return src, nil
+}
+
+// copiedFilesMatch checks to see if the node instruction is a COPY or ADD.
+// If it is either of those two it checks the timestamps on all the files copied/added
+// by the dockerfile. If the host version has a time stamp greater than the time stamp
+// of the build, the build will not use the cached version and will rebuild.
+func (b *Executor) copiedFilesMatch(node *parser.Node, historyTime *time.Time) (bool, error) {
+	if node.Value != "add" && node.Value != "copy" {
+		return true, nil
+	}
+
+	src, err := b.getFilesToCopy(node)
+	if err != nil {
+		return false, err
+	}
+	for _, item := range src {
+		// for urls, check the Last-Modified field in the header.
+		if strings.HasPrefix(item, "http://") || strings.HasPrefix(item, "https://") {
+			urlContentNew, err := urlContentModified(item, historyTime)
+			if err != nil {
+				return false, err
+			}
+			if urlContentNew {
+				return false, nil
+			}
+			continue
+		}
+		// For local files, check the time stamps.
+		timeIsGreater, err := timeStampAfter(item, historyTime)
+		if err != nil {
+			return false, errors.Wrapf(err, "error checking if time stamp is greater than cached version %q", item)
+		}
+		if timeIsGreater {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// urlContentModified sends a get request to the url and checks if the header has a value in
+// Last-Modified, and if it does compares the time stamp to that of the history of the cached image.
+// returns true if there is no Last-Modified value in the header.
+func urlContentModified(url string, historyTime *time.Time) (bool, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return false, errors.Wrapf(err, "error getting %q", url)
+	}
+	if lastModified := resp.Header.Get("Last-Modified"); lastModified != "" {
+		lastModifiedTime, err := time.Parse(time.RFC1123, lastModified)
+		if err != nil {
+			return false, errors.Wrapf(err, "error parsing time for %q", url)
+		}
+		return lastModifiedTime.After(*historyTime), nil
+	}
+	logrus.Debugf("Response header did not have Last-Modified %q, will rebuild.", url)
+	return true, nil
+}
+
+// timeStampAfter returns true if the timestamp of item is greater than historyTime.
+// If item is a directory, it goes through the contents of the sub-directories till
+// a file or directory is found with a time stamp greater than historyTime.
+// If no such case is found, timeStampAfter returns false.
+func timeStampAfter(item string, historyTime *time.Time) (bool, error) {
+	fi, err := os.Stat(item)
+	if err != nil {
+		return false, errors.Wrapf(err, "error stat %q", item)
+	}
+	if fi.ModTime().After(*historyTime) {
+		return true, nil
+	}
+	if fi.Mode().IsDir() {
+		files, err := ioutil.ReadDir(item)
+		if err != nil {
+			return false, errors.Wrapf(err, "error retrieving contents of directory %q", item)
+		}
+		for _, file := range files {
+			return timeStampAfter(filepath.Join(item, file.Name()), historyTime)
```

Comment:
Not sure if using recursion is a good idea, should we just walk the directory?  Ask @nalind

Reply:
Realised that filpath has a Walk functions, so using that instead to walk to file tree and compare the time stamps. filpath.Walk walks the file tree recursively I believe though.